app/views/shared/_error_messages.html.erb

```diff
@@ -0,0 +1,12 @@
+<% if @user.errors.any? %>
+    <div id="error_explanation">
+        <div class="alert alert-danger">
+            The form contains <%= pluralize(@user.errors.count, "error") %>.
+        </div>
+        <ul>
+            <% @user.errors.full_messages.each do |msg| %>
```

Comment:
One fun & instructive thing to do is to explore the structure of `ActiveModel::Errors` in a Rails console (having `pry` installed):
```ruby
(main)> user = User.new
=> #<User:0x00007fe6ddb0de40
 id: nil,
 ....

(main)> user.errors.count
=> 0

# errors only get populated after running validations
(main)> user.valid?
=> false

(main)> cd user.errors
(#<ActiveModel::Errors:0x00007fe6df8786b0>)> ls
Enumerable#methods: 
  all?            each_cons         first     max        reduce        take                                  
  any?            each_entry        flat_map  max_by     reject        take_while                            
  chunk           each_slice        grep      member?    reverse_each  to_h                                  
  chunk_while     each_with_index   grep_v    min        select        to_json                               
  collect         each_with_object  group_by  min_by     slice_after   to_json_with_active_support_encoder   
  collect_concat  entries           index_by  minmax     slice_before  to_json_without_active_support_encoder
  cycle           exclude?          inject    minmax_by  slice_when    to_set                                
  detect          find              lazy      none?      sort          uniq                                  
  drop            find_all          many?     one?       sort_by       zip                                   
  drop_while      find_index        map       partition  sum         
ActiveModel::Errors#methods: 
  []   add_on_blank  as_json  count   empty?         full_messages_for  has_key?  keys      size     to_xml
  []=  add_on_empty  blank?   delete  full_message   generate_message   include?  messages  to_a     values
  add  added?        clear    each    full_messages  get                key?      set       to_hash
self.methods: __pry__
instance variables: @base  @messages
locals: _  __  _dir_  _ex_  _file_  _in_  _out_  _pry_

(#<ActiveModel::Errors:0x00007fe6df8786b0>)> messages
=> {:password=>["can't be blank", "is too short (minimum is 8 characters)", "is invalid"], :username=>["can't be blank"]}

# feel free to explore more stuff
```

One of the most valuable skills you can learn as a Ruby/Rails developer is getting familiar with using `pry`commands to quickly... `pry` out information from data. It's probably my favourite Ruby gem :) Here's an article outlining some of the cool things you can do with it https://medium.com/rubyinside/powering-your-ruby-rails-development-with-pry-3d5dbd2a8b80

Here's a cheatsheet I've also put together for useful things one can do with pry [pry.zip](https://github.com/JOlsen-Neslo/rails-app/files/3690858/pry.zip)

Reply:
awesome comment @c13e

Reply:
@c13e thanks! This is awesome